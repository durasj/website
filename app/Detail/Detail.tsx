import React from 'react';

import ImageGallery from '../../node_modules/react-image-gallery/build/image-gallery';

import { IProject } from '../Projects';

export default class Detail extends React.Component<{project: IProject}, {}> {
    protected static propTypes = {
        project: React.PropTypes.object.isRequired,
    };

    private _imageGallery;
    private images;

    constructor(props: {project: IProject}) {
        super(props);

        if (this.props.project.photos !== undefined) {
            this.images = this.props.project.photos.map(photo => {
                return {
                    original: photo.src,
                    thumbnail: photo.src,
                    originalAlt: photo.caption,
                    thumbnailAlt: photo.caption,
                };
            });
        }
    }

    public render(): React.ReactElement<{project: IProject}> {
        const skillsList = this.props.project.skills ? this.props.project.skills.join(', ') : undefined;
        const period = this.props.project.period ? (
            <div className="period">Made in: {this.props.project.period}</div>
        ) : undefined;
        const skills = skillsList ? (
            <div>
                Skills learned/practised: <span className="skills">{skillsList}</span>
            </div>
        ) : undefined;
        const gallery = this.images !== undefined ? (
            <ImageGallery
                ref={i => this._imageGallery = i}
                items={this.images}
                showNav={false}
                showPlayButton={false}
            />
        ) : undefined;

        return (
            <div id="detail">
                <div className={"section is-" + (this.props.project.color || 'bluegrey')}>
                    <div className="container">
                        <div className="columns is-mobile">
                            <div className="column">
                                <h1 className="title is-3">{this.props.project.title}</h1>
                                <h2 className="subtitle">{this.props.project.description}</h2>
                            </div>
                            <div className="column is-narrow">
                                <span className="close" onClick={this.close.bind(this)}>&times;</span>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="section">
                    <div className="container">
                        <div className="columns">
                            <div
                                className="column content"
                                dangerouslySetInnerHTML={{__html: this.props.project.content}}
                            />
                            <div className="column is-one-quarter meta">
                                {period}

                                {skills}
                            </div>
                        </div>

                        {gallery}
                    </div>
                </div>
            </div>
        );
    }

    protected close() {
        location.hash = '#';
    }
}
