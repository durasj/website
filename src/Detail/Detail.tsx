import { LuminousGallery } from 'luminous-lightbox';
import React, { Component, EventHandler, MouseEvent, ReactElement } from 'react';

import DocumentTitle from 'react-document-title';

import { IProject } from '../Projects';

interface IProps {
    project: IProject,
    onClose: () => void,
}

export default class Detail extends Component<IProps, { closing: boolean }> {
    private handleClosing: EventHandler<MouseEvent<HTMLElement>>;
    private gallery;

    constructor(props: IProps) {
        super(props);

        this.state = { closing: false };

        this.handleClosing = this.close.bind(this);
    }

    public componentDidMount() {
        window.scrollTo(0, 0);

        this.gallery = new LuminousGallery(
            document.querySelectorAll('.gallery a'),
            {},
            {
                caption: (trigger) => (
                    trigger.querySelector('img').getAttribute('alt')
                ),
            },
        );
    }

    public componentWillUnmount() {
        this.gallery.destroy();
    }

    public shouldComponentUpdate(nextProps, nextState) {
        return nextProps !== this.props || nextState !== this.state;
    }

    public render(): ReactElement<{project: IProject}> {
        const documentTitle = this.props.project.title + ' | ' + 'Jakub Duras';
        const skillsList = this.props.project.skills ? (
            this.props.project.skills.map(
                (skill) => <span key={skill} className="skill">{skill}</span>,
            )
        ) : undefined;
        const period = this.props.project.period ? (
            <div className="period">{this.props.project.period}</div>
        ) : undefined;
        const skills = skillsList ? (
            <div className="skills">
                {skillsList}
            </div>
        ) : undefined;
        const photos = this.props.project.photos;
        const gallery = photos !== undefined ? (
            photos.map(
                (photo) => (
                    <a key={photo.src} href={photo.src}>
                        <img src={photo.src} alt={photo.caption} />
                    </a>
                ),
            )
        ) : undefined;
        const link = this.props.project.link ? (
            <a
                href={this.props.project.link}
                className="link button button-outline"
            >{this.props.project.linkLabel}</a>
        ) : undefined;

        return (
            <div className={"detail" + (this.state.closing ? ' closing' : '')}>
                <div className="header">
                    <div>
                        <DocumentTitle title={documentTitle} />
                        <h1 className="title">{this.props.project.title}</h1>
                    </div>

                    <span className="close" onClick={this.handleClosing}>&times;</span>
                </div>
                <div className="body">
                    <div className="gallery">
                        {gallery}
                    </div>

                    <div className="content">
                        <h2>Description</h2>
                        <div
                            className="typography"
                            dangerouslySetInnerHTML={{__html: this.props.project.content}}
                        />
                        {skills}

                        <div className="bottom">
                            {link}
                            {period}
                        </div>
                    </div>
                </div>
            </div>
        );
    }

    private close() {
        this.setState({ closing: true });

        setTimeout(this.props.onClose, 350);
    }
}
