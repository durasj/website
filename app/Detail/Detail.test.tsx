import { mount } from 'enzyme';
import { LuminousGallery } from 'luminous-lightbox';
import React from 'react';
import renderer from 'react-test-renderer';

import mockProjects from '../../testing/mockProjects';
import Detail from '../Detail';

beforeEach(() => {
    LuminousGallery.mockClear();
});

it('Renders with all info', () => {
    const project = mockProjects[0];
    window.scrollTo = jest.fn();

    const component = renderer.create(
        <Detail project={project} onClose={() => {}} />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('Renders without photos, skills and period', () => {
    const project = {
        ...mockProjects[0],
        skills: undefined,
        period: undefined,
        photos: undefined,
    };
    window.scrollTo = jest.fn();

    const component = renderer.create(
        <Detail project={project} onClose={() => {}} />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('Scrolls up on mount', () => {
    const project = mockProjects[0];
    window.scrollTo = jest.fn();

    renderer.create(
        <Detail project={project} onClose={() => {}} />,
    );

    expect(window.scrollTo).toHaveBeenCalledWith(0, 0);
});

it('Calls onClose when closed', () => {
    const project = mockProjects[0];
    const handleClosing = jest.fn();
    window.scrollTo = jest.fn();

    const component = mount(
        <Detail project={project} onClose={handleClosing} />,
    );

    jest.useFakeTimers();
    component.find('.close').simulate('click', { preventDefault: () => undefined });
    jest.runAllTimers();

    expect(handleClosing).toHaveBeenCalled();
});

it('Creates and destroys lightbox gallery', () => {
    const project = mockProjects[0];
    window.scrollTo = jest.fn();
    const component = mount(
        <Detail project={project} onClose={() => {}} />,
    );

    expect(LuminousGallery.mock.instances.length).toBe(1);
    const mockGalleryInstance = LuminousGallery.mock.instances[0];

    component.unmount();

    expect(mockGalleryInstance.destroy).toHaveBeenCalled();
});
