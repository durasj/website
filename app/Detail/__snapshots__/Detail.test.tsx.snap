// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with all info 1`] = `
<div
  className="detail"
>
  <div
    className="header"
  >
    <div>
      <h1
        className="title"
      >
        Something
      </h1>
    </div>
    <span
      className="close"
      onClick={[Function]}
    >
      ×
    </span>
  </div>
  <div
    className="body"
  >
    <div
      className="gallery"
    >
      <a
        href="/google/random.jpg"
      >
        <img
          alt="Random photo I found on Google"
          src="/google/random.jpg"
        />
      </a>
      <a
        href="/google/funny.jpg"
      >
        <img
          alt="Funny image from Google"
          src="/google/funny.jpg"
        />
      </a>
    </div>
    <div
      className="content"
    >
      <h2>
        Description
      </h2>
      <div
        className="typography"
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
      />
      <div
        className="skills"
      >
        <span
          className="skill"
        >
          all
        </span>
        <span
          className="skill"
        >
          right
        </span>
        <span
          className="skill"
        >
          skills
        </span>
      </div>
      <div
        className="bottom"
      >
        <a
          className="link button button-outline"
          href="https://google.com"
        >
          My weekend project
        </a>
        <div
          className="period"
        >
          2222
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders without photos, skills and period 1`] = `
<div
  className="detail"
>
  <div
    className="header"
  >
    <div>
      <h1
        className="title"
      >
        Something
      </h1>
    </div>
    <span
      className="close"
      onClick={[Function]}
    >
      ×
    </span>
  </div>
  <div
    className="body"
  >
    <div
      className="gallery"
    />
    <div
      className="content"
    >
      <h2>
        Description
      </h2>
      <div
        className="typography"
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
      />
      <div
        className="bottom"
      >
        <a
          className="link button button-outline"
          href="https://google.com"
        >
          My weekend project
        </a>
      </div>
    </div>
  </div>
</div>
`;
